File: AdventOfCode/3/2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>

using namespace std;

// Function to read the input from a file
string readInput(const string& filePath) {
    ifstream file(filePath);
    if (!file.is_open()) {
        cerr << "Error: Unable to open file " << filePath << endl;
        exit(1);
    }

    string content((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
    file.close();
    return content;
}

// Function to process instructions and calculate the sum
int processInstructions(const string& data) {
    // Regular expressions for instructions
    regex mulPattern(R"(mul\((\d+),(\d+)\))");
    regex doPattern(R"(do\(\))");
    regex dontPattern(R"(don't\(\))");

    // State and result initialization
    bool isEnabled = true;
    int totalSum = 0;

    // Combined regex for all patterns
    regex combinedPattern(R"(mul\(\d+,\d+\)|do\(\)|don't\(\))");
    auto wordsBegin = sregex_iterator(data.begin(), data.end(), combinedPattern);
    auto wordsEnd = sregex_iterator();

    for (auto it = wordsBegin; it != wordsEnd; ++it) {
        string instruction = it->str();

        if (regex_match(instruction, doPattern)) {
            isEnabled = true;
        } else if (regex_match(instruction, dontPattern)) {
            isEnabled = false;
        } else if (regex_match(instruction, mulPattern)) {
            if (isEnabled) {
                smatch matches;
                regex_search(instruction, matches, mulPattern);

                int x = stoi(matches[1].str());
                int y = stoi(matches[2].str());

                totalSum += x * y;
            }
        }
    }

    return totalSum;
}

int main() {
    // Path to the input file
    string filePath = "input.txt";

    // Read and process the input
    string data = readInput(filePath);
    int result = processInstructions(data);

    // Output the result
    cout << "Sum of enabled multiplications: " << result << endl;

    return 0;
}

File: AdventOfCode/3/2.py
import re

def parse_input(file_path):
    with open(file_path, 'r') as file:
        data = file.read()
    return data

def process_instructions(data):
    # Regular expressions to match instructions
    mul_pattern = re.compile(r'mul\((\d+),(\d+)\)')
    do_pattern = re.compile(r'do\(\)')
    dont_pattern = re.compile(r"don't\(\)")

    # Initialize the state and result
    is_enabled = True
    total_sum = 0

    # Iterate through the matches
    for match in re.finditer(r'mul\(\d+,\d+\)|do\(\)|don\'t\(\)', data):
        instruction = match.group(0)

        if instruction == 'do()':
            is_enabled = True
        elif instruction == "don't()":
            is_enabled = False
        else:
            # It's a mul() instruction
            if is_enabled:
                x, y = map(int, mul_pattern.match(instruction).groups())
                total_sum += x * y

    return total_sum

def main():
    # Path to the input file
    file_path = '/Users/aditya/Documents/CP/AdventOfCode/3/input.txt'
    data = parse_input(file_path)
    result = process_instructions(data)
    print(f"Sum of enabled multiplications: {result}")

if __name__ == "__main__":
    main()





# Total sum of enabled mul instructions: 178794710